### PR TITLE
MDT-69, MDT-102 As a doctor, I want to be able to create  a user for a patient

### PR DESCRIPTION
**Pull request by TestaDiRapa**

### CODE REVIEW

File: src/apis/AuthenticationApi.ts [hunk 1]
```diff
@@ -46,4 +46,23 @@ export interface AuthenticationApi {
...
+   * - Send a Notification to all the delegated HCP to ask for access to the data of the Patient
+   * @param userLogin The login of the user
+   * @param shortLivedToken The short-lived authentication token created by the HCP
+   * @param userKeyPair The key pair [private, public] that will be used by the user to encrypt/decrypt data;
+   * @param tokenAndKeyPairProvider A custom function to generate an authentication token and a key pair for user
+   *
+   * @return The result of the authentication and the related MedTechApi object corresponding to the newly authenticated
+   * user.
+   */
+  authenticateAndAskForAccess(
```
Comment (lisatr2812) [inline src/apis/AuthenticationApi.ts:62]: Rename to `authenticateAndAskAccessToItsExistingData`

File: src/apis/NotificationApi.ts [hunk 2]
```diff
@@ -33,5 +33,19 @@ export interface NotificationApi {
...
    * @return a Promise containing the Notification or undefined if something goes wrong.
    */
-  getNotification(notificationId: string): Promise<Notification | undefined>
+  getNotification(notificationId: string): Promise<Notification | undefined>;
+
+  /**
+   * Gets all the Notifications with type NEW_USER_OWN_DATA_ACCESS and status "pending" that the current dataOwner can access
+   * @return an Array of the Notifications matching those criteria
+   */
+  getPendingNotificationsFromNewUsers(): Promise<Array<Notification>>;
```
Comment (lisatr2812) [inline src/apis/NotificationApi.ts:42]: You can rename in `getPendingNotifications` and return all types of them. The user will treat them one by one afterwards.

File: src/apis/impl/AuthenticationApiImpl.ts [hunk 3]
```diff
@@ -100,10 +102,69 @@ export class AuthenticationApiImpl implements AuthenticationApi {
     return null
   }
 
-  async initApiAndUserAuthenticationToken(process: AuthenticationProcess, validationCode: string, tokenAndKeyPairProvider: (groupId: string, userId: string) => ([string, [string, string]] | undefined)): Promise<[MedTechApi, ApiInitialisationResult]> {
+  async authenticateAndAskForAccess(userLogin: string, shortLivedToken: string, dataOwnerKeyPair: [string, string] | undefined, tokenAndKeyPairProvider: (groupId: string, userId: string) => ([string, [string, string]] | undefined)): Promise<AuthenticationResult | null> {
+    const authenticationResult = await this.initUserTokenAndCrypto(userLogin, shortLivedToken, dataOwnerKeyPair, tokenAndKeyPairProvider);
+
+    const loggedUser = await authenticationResult.medTechApi.userApi.getLoggedUser();
+    if (!loggedUser) throw new Error("User log in failed");
+    if (!loggedUser.patientId) throw new Error("User is not associated to any Patient");
```
Comment (lisatr2812) [inline src/apis/impl/AuthenticationApiImpl.ts:110]: I would be more generic and check if the loggedUser has a dataOwner instead of a patient only : That way, if the developer would like to createAndInvite another healthcareParty, he can use both services (createAndInvite & authenticateAndAskAccessToItsExistingData)

File: src/apis/impl/AuthenticationApiImpl.ts [hunk 4]
```diff
@@ -100,10 +102,69 @@ export class AuthenticationApiImpl implements AuthenticationApi {
...
+
+    const patientUser = await authenticationResult.medTechApi.patientApi.getPatient(loggedUser.patientId);
```
Comment (lisatr2812) [inline src/apis/impl/AuthenticationApiImpl.ts:112]: Same : You have a getDataOwner API you can use in the base sdk to treat patient / hcp & device. That way, the service will be working for each type of user

File: src/apis/impl/AuthenticationApiImpl.ts [hunk 5]
```diff
@@ -100,10 +102,69 @@ export class AuthenticationApiImpl implements AuthenticationApi {
...
+    if (!patientUser) throw new Error(`Patient with id ${loggedUser.patientId} does not exist`);
+
+    const delegates = Object.entries(patientUser.systemMetaData?.delegations ?? {})
+      .map( (keyValue) => Array.from(keyValue[1]))
+      .flat()
+      .filter( (delegation) => !!delegation.delegatedTo)
+      .map( (delegation) => delegation.delegatedTo!);
+
+    const hcpDelegates = await authenticationResult.medTechApi.healthcareProfessionalApi.filterHealthcareProfessionalBy(
```
Comment (lisatr2812) [inline src/apis/impl/AuthenticationApiImpl.ts:121]: You don't need the `hcpDelegates` variable : the hcp.Id is the same than the delegation.delegatedFrom. 
You can therefore directly use the list of delegatedFrom & delegatedTo in the patient. 
For Patient PATIENT_1, created by HCP_1, you take the delegation ids in bold : 
- delegatedFrom : **HCP_1** - delegatedTo : PATIENT_1;
- delegatedFrom : HCP_1 - delegatedTo : **HCP_2**;
- delegatedFrom : HCP_1 - delegatedTo : **HCP_3**;
- delegatedFrom : HCP_2 - delegatedTo : **HCP_4**;

File: src/apis/impl/DataSampleApiImpl.ts [hunk 6]
```diff
@@ -863,6 +865,35 @@ export class DataSampleApiImpl implements DataSampleApi {
...
+  async concatenateFilterResults(filter: Filter<DataSample>, nextId?: string | undefined, limit?: number | undefined, accumulator: Array<DataSample> = []): Promise<Array<DataSample>> {
+    const paginatedDataSamples = await this.filterDataSample(filter, nextId, limit);
+    return !paginatedDataSamples.nextKeyPair?.startKeyDocId
+      ? accumulator.concat(paginatedDataSamples.rows)
+      : this.concatenateFilterResults(filter, paginatedDataSamples.nextKeyPair.startKeyDocId, limit, accumulator.concat(paginatedDataSamples.rows))
+  }
+
+  async getDataSamplesForPatient(patient: Patient): Promise<Array<DataSample>> {
+    const user = await this.userApi.getCurrentUser();
+    if (!user) throw new Error("There is no user currently logged in");
```
Comment (lisatr2812) [inline src/apis/impl/DataSampleApiImpl.ts:887]:
```
if (!user) {
   throw ...
}
```

File: src/apis/impl/DataSampleApiImpl.ts [hunk 7]
```diff
@@ -863,6 +865,35 @@ export class DataSampleApiImpl implements DataSampleApi {
...
+    const dataOwnerId = user.healthcarePartyId ?? user.patientId ?? user.deviceId;
```
Comment (lisatr2812) [inline src/apis/impl/DataSampleApiImpl.ts:888]: `const dataOwnerId = this.userApi.getDataOwnerOf(user)`

File: src/apis/impl/DataSampleApiImpl.ts [hunk 8]
```diff
@@ -863,6 +865,35 @@ export class DataSampleApiImpl implements DataSampleApi {
...
+    if (!dataOwnerId) throw new Error("User is not a Data Owner");
```
Comment (lisatr2812) [inline src/apis/impl/DataSampleApiImpl.ts:889]:
```
if (!dataOwnerId) {
   throw ...
}
```

File: src/apis/impl/DataSampleApiImpl.ts [hunk 9]
```diff
@@ -863,6 +865,35 @@ export class DataSampleApiImpl implements DataSampleApi {
       })
   }
 
+  async giveAccessToMany(dataSamples: Array<DataSample>, delegatedTo: string): Promise<Array<string>> {
+    return (await Promise.all(dataSamples.map(async dataSample => {
+      try {
+        return (await this.giveAccessTo(dataSample, delegatedTo)).id
```
Comment (lisatr2812) [inline src/apis/impl/DataSampleApiImpl.ts:871]: This is an operation that's gonna take a lot of time which implies two things to do : 
- Try to avoid any additional requests. I would for example save all contacts at the same time, instead of treating them one by one. 
- Add a service `mayAccess(dataSample, dataOwnerId)`, so that if a problem happens during the treatment (app shutdown, ...) the user will be able to continue its treatment easily, taking back the ones that haven't been treated. 
We need to think about this point together, let's discuss it

File: src/apis/impl/HealthcareElementApiImpl.ts [hunk 10]
```diff
@@ -214,4 +216,34 @@ export class HealthcareElementApiImpl implements HealthcareElementApi {
         return HealthcareElementMapper.toHealthcareElement(updatedHealthElement)!
       })
   }
+
+  async giveAccessToMany(healthElements: Array<HealthcareElement>, delegatedTo: string): Promise<Array<string>> {
```
Comment (lisatr2812) [inline src/apis/impl/HealthcareElementApiImpl.ts:220]: We need to think deeper about the giveAccessToMany, let's discuss about it together

File: src/apis/impl/HealthcareElementApiImpl.ts [hunk 11]
```diff
@@ -214,4 +216,34 @@ export class HealthcareElementApiImpl implements HealthcareElementApi {
...
+  async concatenateFilterResults(filter: Filter<HealthcareElement>, nextId?: string | undefined, limit?: number | undefined, accumulator: Array<HealthcareElement> = []): Promise<Array<HealthcareElement>> {
+    const paginatedHealthcareElements = await this.filterHealthcareElement(filter, nextId, limit);
+    return !paginatedHealthcareElements.nextKeyPair?.startKeyDocId
+      ? accumulator.concat(paginatedHealthcareElements.rows)
+      : this.concatenateFilterResults(filter, paginatedHealthcareElements.nextKeyPair.startKeyDocId, limit, accumulator.concat(paginatedHealthcareElements.rows))
+  }
+
+  async getHealthcareElementsForPatient(patient: Patient): Promise<Array<HealthcareElement>> {
+    const user = await this.userApi.getCurrentUser();
+    if (!user) throw new Error("There is no user currently logged in");
```
Comment (lisatr2812) [inline src/apis/impl/HealthcareElementApiImpl.ts:239]:
```
if (!user) {
   throw ...
}
```

File: src/apis/impl/HealthcareElementApiImpl.ts [hunk 12]
```diff
@@ -214,4 +216,34 @@ export class HealthcareElementApiImpl implements HealthcareElementApi {
...
+    const dataOwnerId = user.healthcarePartyId ?? user.patientId ?? user.deviceId;
```
Comment (lisatr2812) [inline src/apis/impl/HealthcareElementApiImpl.ts:240]: `const dataOwnerId = this.userApi.dataOwnerOf(user)`

File: src/apis/impl/HealthcareElementApiImpl.ts [hunk 13]
```diff
@@ -214,4 +216,34 @@ export class HealthcareElementApiImpl implements HealthcareElementApi {
...
+    if (!dataOwnerId) throw new Error("User is not a Data Owner");
```
Comment (lisatr2812) [inline src/apis/impl/HealthcareElementApiImpl.ts:241]:
```
if (!dataOwnerId) {
   throw ...
}
```

File: src/apis/impl/NotificationApiImpl.ts [hunk 14]
```diff
@@ -97,4 +94,27 @@ export class NotificationApiImpl implements NotificationApi {
...
 
+  async concatenateFilterResults(filter: Filter<Notification>, nextId?: string | undefined, limit?: number | undefined, accumulator: Array<Notification> = []): Promise<Array<Notification>> {
+    const paginatedNotifications = await this.filterNotifications(filter, nextId, limit);
+    return !paginatedNotifications.nextKeyPair?.startKeyDocId
+      ? accumulator.concat(paginatedNotifications.rows)
+      : this.concatenateFilterResults(filter, paginatedNotifications.nextKeyPair.startKeyDocId, limit, accumulator.concat(paginatedNotifications.rows))
+  }
+
+  async getPendingNotificationsFromNewUsers(): Promise<Array<Notification>> {
+    const user = await this.userApi.getCurrentUser();
```
Comment (lisatr2812) [inline src/apis/impl/NotificationApiImpl.ts:105]:
```
if (!user) {
   throw ...
}
```

File: src/apis/impl/NotificationApiImpl.ts [hunk 15]
```diff
@@ -97,4 +94,27 @@ export class NotificationApiImpl implements NotificationApi {
...
+    if (!user) throw new Error("There is no user currently logged in");
+    if (!user.healthcarePartyId) throw new Error("User is not a Healthcare Party");
```
Comment (lisatr2812) [inline src/apis/impl/NotificationApiImpl.ts:107]:
```
if (!user.healthcarePartyId) {
   throw ...
}
```

Comment (lisatr2812) [inline src/apis/impl/NotificationApiImpl.ts:107]: Check if the user is a dataOwner instead of a HCP : A patient can also receive notifications

File: src/apis/impl/NotificationApiImpl.ts [hunk 16]
```diff
@@ -97,4 +94,27 @@ export class NotificationApiImpl implements NotificationApi {
...
+    const filter = await new NotificationFilter()
+      .forHcParty(user.healthcarePartyId)
+      .withType(NotificationTypeEnum.NEW_USER_OWN_DATA_ACCESS)
```
Comment (lisatr2812) [inline src/apis/impl/NotificationApiImpl.ts:110]: Take all type of notifications for the current user, he will treat them one by one afterwards

File: src/filter/filterDsl.ts [hunk 17]
```diff
@@ -357,9 +357,12 @@ export class HealthcareElementFilter implements FilterBuilder<HealthcareElement>
         patientSecretForeignKeys: (await Promise.all(
           this._forPatients[1].map(async (p) =>
             (await this._forPatients![0].extractKeysHierarchyFromDelegationLikes(dataOwnerId, p.id!, Object.entries(p.systemMetaData!.delegations!)
-              .map(([k, v]) => [k, Array.from(v)] as [string, Delegation[]]).reduce((m, [k, v]) => {m[k] = v; return m}, {} as {[key: string]: Delegation[]})
+              .map(([k, v]) => [k, Array.from(v)] as [string, Delegation[]])
+              .reduce((m, [k, v]) => {
+                m[k] = v; return m}, {} as {[key: string]: Delegation[]
+              })
             )).map((x) => x.extractedKeys))
-        )).reduce((t,v) => t.concat(v[1]) ,[] as string[])
```
Comment (lisatr2812) [inline src/filter/filterDsl.ts:362]: Why did you change the delegation id index to take ?

File: src/utils/gatewayMessageFactory.ts [hunk 18]
```diff
@@ -0,0 +1,77 @@
+import {HealthcareProfessional} from "../models/HealthcareProfessional";
+import {Patient} from "../models/Patient";
+import {User} from "../models/User";
+import {filteredContactsFromAddresses} from "./addressUtils";
+
+export type SMSMessage = {message: string};
+export type EmailMessage = {from: string, html: string, subject: string}
+
+export interface GatewayMessageFactory<T> {
```
Comment (lisatr2812) [inline src/utils/gatewayMessageFactory.ts:9]: MsgGtwMessageFactory

File: src/utils/gatewayMessageFactory.ts [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+
+export class ICureRegistrationEmail implements EmailMessageFactory {
+  hcp: HealthcareProfessional;
+  hcpEmail: string;
+  link: string;
+  solutionName: string;
+  patient: Patient;
+
+  constructor(
+    hcp: HealthcareProfessional,
```
Comment (lisatr2812) [inline src/utils/gatewayMessageFactory.ts:29]: Take a dataOwner instead of a HCP only

**Review by lisatr2812 — CHANGES_REQUESTED** [review]

File: src/utils/gatewayMessageFactory.ts (hunk 19 above)
Comment (TestaDiRapa) [reply src/utils/gatewayMessageFactory.ts:29]: Currently there is no dataOwner model, should I create it?

File: src/apis/impl/NotificationApiImpl.ts (hunk 16 above)
Comment (TestaDiRapa) [reply src/apis/impl/NotificationApiImpl.ts]: We have a problem for this one, as there isn't a Filter to get all the notification for a single HCP, so if we remove the type we will get all the notifications for all the HCP

File: src/apis/impl/AuthenticationApiImpl.ts (hunk 5 above)
Comment (lisatr2812) [reply src/apis/impl/AuthenticationApiImpl.ts]: It could be a possibility in the future (I want to share this data with my wife/husband, ...)

File: src/apis/impl/NotificationApiImpl.ts (hunk 16 above)
Comment (lisatr2812) [reply src/apis/impl/NotificationApiImpl.ts]: True : I'm adding a healthcarePartyId field in the afterDate filter (which is even mandatory, to avoid users to have notifications of their whole group) so that we can have a quick fix for this

File: src/utils/gatewayMessageFactory.ts (hunk 19 above)
Comment (lisatr2812) [reply src/utils/gatewayMessageFactory.ts]: No, let's use Typescript Multiple Type Definition : In the future, we'll have a service and a DTO dedicated to the dataOwner in the base SDK, let's not spend time for it today.
